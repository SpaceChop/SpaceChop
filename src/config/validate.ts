import console from '../lib/console';
import validate from '../types/validate';

let valid = true;

export default (config, throwError = false) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.info('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.error('\nThere is an error in /config.yml:');
    console.error('--------------------------------');
    console.error(err.message);
    console.error('--------------------------------');
    if (throwError) {
      throw new Error(err);
    }
  }
  return false;
};
