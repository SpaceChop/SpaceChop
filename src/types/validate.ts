import Config from './Config';

export default (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key ? `${err.key}: ` : ''}${err.message}`);
  }
};
