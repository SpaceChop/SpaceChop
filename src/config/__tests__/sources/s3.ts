import { validate } from '../../validate';

describe('validate s3 source', () => {
  it('should throw on invalid source', () => {
    const config = {
      paths: ['/:preset/:image'],
      sources: [{ s3: {} }],
      presets: {},
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should throw on invalid params in source', () => {
    const config = {
      paths: ['/:preset/:image'],
      sources: [{ s3: {
      access_key_id: '',
      secret_access_key: '',
      region: '',
      bucket_name: '',
      path: '',
    } }], presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should not throw on valid source', () => {
    const config = {
      paths: ['/:preset/:image'],
      sources: [{
        s3: {
          access_key_id: 'access_key_id',
          secret_access_key: 'secret_access_key',
          region: 'region',
          bucket_name: 'bucket_name',
          path: ':image',
        },
      }],
      presets: {},
    };
    expect(() => validate(config)).not.toThrow();
  });
});
