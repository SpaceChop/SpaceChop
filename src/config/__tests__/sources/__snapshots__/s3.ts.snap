// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate s3 source should throw on empty path in source 1`] = `
"source[0].s3.path: Path cannot be empty
params.image: Could not find any usage in sources, storage or paths"
`;

exports[`validate s3 source should throw on invalid params in source 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; } & { endpoint?: string; }; } | { volume: { root: string; }; }, but was object"`;

exports[`validate s3 source should throw on invalid source 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; } & { endpoint?: string; }; } | { volume: { root: string; }; }, but was object"`;

exports[`validate s3 source should throw on missing params in path on source 1`] = `
"source[0].s3.path: Missing param \\"image\\" in paths
params.uuid: Could not find any usage in sources, storage or paths"
`;

exports[`validate s3 source should throw on missing params in source 1`] = `
"storage.s3.path: Missing path parameters
params.image: Could not find any usage in sources, storage or paths"
`;
